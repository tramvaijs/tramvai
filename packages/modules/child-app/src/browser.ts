import './shared/typings';

import { Module } from '@tramvai/core';
import { sharedProviders } from './shared/providers';
import { browserProviders } from './browser/providers';

export * from './export';

@Module({
  imports: [],
  providers: [...sharedProviders, ...browserProviders],
})
export class ChildAppModule {}
