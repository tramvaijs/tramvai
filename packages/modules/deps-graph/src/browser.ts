import { Module } from '@tramvai/core';

@Module({
  providers: [],
})
export class DepsGraphModule {}
