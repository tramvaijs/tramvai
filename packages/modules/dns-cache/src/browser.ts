import { declareModule } from '@tramvai/core';

export const TramvaiDnsCacheModule = declareModule({
  name: 'TramvaiDnsCacheModule',
  imports: [],
  providers: [],
});
