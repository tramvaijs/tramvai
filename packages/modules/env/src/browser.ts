import { Module, Scope } from '@tramvai/core';
import {
  STORE_TOKEN,
  COMBINE_REDUCERS,
  ENV_MANAGER_TOKEN,
  ENV_USED_TOKEN,
} from '@tramvai/tokens-common';

import { EnvironmentManager } from './shared/EnvironmentManager';
import { EnvironmentStore } from './shared/EnvironmentStore';

export { ENV_MANAGER_TOKEN, ENV_USED_TOKEN };

@Module({
  providers: [
    {
      provide: ENV_MANAGER_TOKEN,
      useFactory: ({ store }) => {
        const env = new EnvironmentManager();
        env.update(store.getState(EnvironmentStore));

        return env;
      },
      scope: Scope.SINGLETON,
      deps: {
        store: STORE_TOKEN,
      },
    },
    {
      provide: COMBINE_REDUCERS,
      useValue: EnvironmentStore,
      multi: true,
    },
  ],
})
export class EnvironmentModule {}
