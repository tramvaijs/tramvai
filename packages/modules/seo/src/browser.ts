import flatten from '@tinkoff/utils/array/flatten';
import isArray from '@tinkoff/utils/is/array';
import { Module, Scope, provide } from '@tramvai/core';
import { Meta, Update } from '@tinkoff/meta-tags-generate';
import { COMMAND_LINE_EXECUTION_END_TOKEN } from '@tramvai/tokens-core-private';
import { INITIAL_APP_STATE_TOKEN } from '@tramvai/module-common';
import { transformValue } from './transformValue';
import { sharedProviders } from './shared';
import { converters } from './converters/converters';
import { META_UPDATER_TOKEN, META_DEFAULT_TOKEN, META_WALK_TOKEN } from './tokens';
import type { MetaRouteConfig, SeoModuleOptions, PageSeoProperty } from './types';

export * from './constants';
export * from './tokens';

declare module '@tramvai/react' {
  export interface PageComponentOptions {
    seo?: PageSeoProperty;
  }
}

declare module '@tinkoff/router' {
  export interface RouteConfig {
    meta?: MetaRouteConfig;
  }
}

@Module({
  providers: [
    ...sharedProviders,
    provide({
      // Update meta only when all actions were completed.
      provide: COMMAND_LINE_EXECUTION_END_TOKEN,
      scope: Scope.SINGLETON,
      multi: true,
      useFactory: () => {
        return function updateMeta(di, type, status) {
          if (type === 'client' && (status === 'afterSpa' || status === 'customer')) {
            const metaWalk = di.get(META_WALK_TOKEN);

            // Be careful: because it executes on customer start
            // it can override some metadata that was received(executed) on the client before.
            // We can easily control it by priority value and give the server meta-updates lower priority in these scenarios.
            if (status === 'customer') {
              const initialState = di.get({ token: INITIAL_APP_STATE_TOKEN, optional: true });
              metaWalk.mergeValuesFromSerializableState(
                initialState.stores.appMeta.serverMetaWalkState
              );
            }

            // We can't use dependencies below as factory provider dependencies
            // due to dependency cycle when using `@tramvai-tinkoff/module-router`.
            const meta = new Meta({
              list: flatten(
                di.get({ token: META_UPDATER_TOKEN, optional: true, multi: true }) || []
              ),
              metaWalk,
              transformValue,
              converters,
            });

            new Update(meta).update();
          }
        };
      },
    }),
  ],
})
export class SeoModule {
  static forRoot(options: SeoModuleOptions) {
    // легаси ветка для старого формата
    // TODO: убрать
    if (isArray(options)) {
      return {
        mainModule: SeoModule,
        providers: [
          {
            provide: 'metaList',
            useValue: options,
            multi: true,
          },
        ],
      };
    }

    const { metaUpdaters, metaDefault } = options;
    const providers = [];

    if (metaDefault) {
      providers.push({
        provide: META_DEFAULT_TOKEN,
        useValue: metaDefault,
      });
    }

    if (metaUpdaters) {
      providers.push({
        provide: META_UPDATER_TOKEN,
        useValue: metaUpdaters,
        multi: true,
      });
    }

    return {
      providers,
      mainModule: SeoModule,
    };
  }
}
