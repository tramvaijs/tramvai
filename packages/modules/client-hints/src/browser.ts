import { declareModule } from '@tramvai/core';
import { browserProviders } from './shared/providers.browser';

export * from './tokens';
export * from './shared/stores/mediaCheckers';
export * from './shared/stores/mediaSelectors';
export * from './shared/stores/media';
export * from './shared/stores/userAgent';

export { ClientHintsChildAppModule } from './modules/child-app/module';
export { ClientHintsCSRModule } from './modules/csr/browser';

export const ClientHintsModule = /* @__PURE__ */ declareModule({
  name: 'ClientHintsModule',
  providers: browserProviders,
});
