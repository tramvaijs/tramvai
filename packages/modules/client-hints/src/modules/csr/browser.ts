import { declareModule, provide } from '@tinkoff/dippy';
import { STORE_TOKEN, ENV_MANAGER_TOKEN } from '@tramvai/tokens-common';

import { browserProviders } from '../../shared/providers.browser';
import { setUserAgent, UserAgentStore } from '../../shared/stores/userAgent';
import { loadUserAgent } from '../../browser/loadUserAgent';
import { USER_AGENT_TOKEN } from '../../tokens';

declare global {
  interface Window {
    __TRAMVAI_USER_AGENT_DATA?: UADataValues;
  }
}

export const ClientHintsCSRModule = /* @__PURE__ */ declareModule({
  name: 'ClientHintsCSRModule',
  providers: [
    ...browserProviders,
    provide({
      provide: USER_AGENT_TOKEN,
      useFactory: ({ store, envManager }) => {
        if (envManager.get('TRAMVAI_FORCE_CLIENT_SIDE_RENDERING') === 'true') {
          const userAgent = loadUserAgent();

          store.dispatch(setUserAgent(userAgent));

          return userAgent;
        }

        return store.getState(UserAgentStore);
      },
      deps: {
        store: STORE_TOKEN,
        envManager: ENV_MANAGER_TOKEN,
      },
    }),
  ],
});
