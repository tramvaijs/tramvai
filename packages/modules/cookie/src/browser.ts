import { Module, provide, Scope } from '@tramvai/core';
import { ClientHintsModule, USER_AGENT_TOKEN } from '@tramvai/module-client-hints';
import { CookieManager } from './cookieManager.browser';
import { COOKIE_MANAGER_TOKEN } from './tokens';

export { COOKIE_MANAGER_TOKEN };

@Module({
  imports: [ClientHintsModule],
  providers: [
    provide({
      // Управление куками в приложении
      provide: COOKIE_MANAGER_TOKEN,
      useClass: CookieManager,
      scope: Scope.SINGLETON,
      deps: {
        // cookieOptions позволяет переопределить SameSite=Lax на None для фикса проблем с iframe
        // TODO после успешного тестирования в приложении paymentsext, убрать прокидывание cookieOptions
        //  и установить SameSite=None по умолчанию
        cookieOptions: {
          token: 'temporary cookieOptions',
          optional: true,
        },
        userAgent: USER_AGENT_TOKEN,
      },
    }),
  ],
})
export class CookieModule {}
