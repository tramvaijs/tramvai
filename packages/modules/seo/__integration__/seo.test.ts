import { testApp } from '@tramvai/internal-test-utils/testApp';
import { testAppInBrowser } from '@tramvai/internal-test-utils/browser';
import { sleep } from '@tramvai/test-integration';

describe('seo', () => {
  const { getApp } = testApp({
    name: 'seo',
  });

  describe('ssr', () => {
    const getHeadHtml = async (url: string) => {
      const app = getApp();
      const { head } = await app.render(url);

      return head.replace(new RegExp(app.staticUrl, 'g'), 'http://localhost:4000');
    };

    it('test default meta pack', async () => {
      expect(await getHeadHtml('/seo/default/')).toMatchSnapshot();
    });

    it('test common meta', async () => {
      expect(await getHeadHtml('/seo/common/')).toMatchSnapshot();
    });

    it('test twitter meta', async () => {
      expect(await getHeadHtml('/seo/twitter/')).toMatchSnapshot();
    });

    it('test og meta', async () => {
      expect(await getHeadHtml('/seo/og/')).toMatchSnapshot();
    });

    it('test json-ld meta', async () => {
      const { parsed } = await getApp().render('/seo/json-ld/', {
        parserOptions: { blockTextElements: { script: true, style: false } },
      });
      const jsonLdScript = parsed.querySelector('[type="application/ld+json"]');

      expect(jsonLdScript.innerHTML).toMatchSnapshot();
    });
  });

  describe('browser', () => {
    const { getPageWrapper } = testAppInBrowser(getApp);

    it('spa navigations should update meta tags', async () => {
      const { page, router } = await getPageWrapper('/seo/common/');
      expect(await page.title()).toBe('common seo');
      await router.navigate('../twitter');
      expect(await page.title()).toBe('twitter seo');
      await router.navigate('../og');
      expect(await page.title()).toBe('og seo');
    });

    it('should allow to update meta in page actions', async () => {
      const { page, router } = await getPageWrapper('/seo/dynamic/');
      expect(await page.title()).toBe('WoW, such dynamic!');
      await router.navigate('../common/');
      expect(await page.title()).toBe('common seo');
      await router.navigate('../dynamic/');
      await sleep(300);
      expect(await page.title()).toBe('WoW, such dynamic!');
    });

    it('should allow to update meta in browser page actions on first render', async () => {
      const { page, router } = await getPageWrapper('/seo/dynamic-client/');

      await sleep(300);
      expect(await page.title()).toBe('WoW, such dynamic!');

      await router.navigate('../common/');
      expect(await page.title()).toBe('common seo');

      await router.navigate('../dynamic-client/');
      await sleep(300);
      expect(await page.title()).toBe('WoW, such dynamic!');
    });

    it('should have title which is came from the server action', async () => {
      const { page, router } = await getPageWrapper('/seo/dynamic-server/');

      expect(await page.title()).toBe('Hello, this is Tramvai!');
    });

    it('should have title which is came from the server action and then be changed by the action on client', async () => {
      const { page, router } = await getPageWrapper('/seo/dynamic-server-dynamic-client');

      expect(await page.title()).toBe('Hello, this is Tramvai!');

      await page.waitForFunction(
        (expectedTitle) => document.title === expectedTitle,
        'WoW, such dynamic!'
      );
    });
  });
});
