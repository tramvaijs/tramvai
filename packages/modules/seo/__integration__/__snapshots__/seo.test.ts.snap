// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seo ssr test common meta 1`] = `
"
    <meta charset="UTF-8" >
    
    <meta name="attrFromPageService" content="\${SERVER_URL}/seo/common/" data-meta-dynamic="true" >
    <link rel="canonical" href="test-canonical" data-meta-dynamic="true" >
    <meta name="keywords" content="test,common,seo" data-meta-dynamic="true" >
    <meta name="robots" content="none" data-meta-dynamic="true" >
    <meta name="robots" content="noindex, nofollow" data-meta-dynamic="true" >
    <meta name="robots" content="noarchive" data-meta-dynamic="true" >
    <title data-meta-dynamic="true">common seo</title>
    <meta name="viewport" content="test seo" data-meta-dynamic="true" >
    <meta name="metaCustomNameAttribute" content="metaCustomContent" data-meta-dynamic="true" >
    

    
    <script></script>
    

    
    
    <style type="text/css"></style>
    <link crossorigin="anonymous" data-href="\${STATIC_URL}/dist/client/root.chunk.css" rel="stylesheet" >
    

    

    

    

    

    
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/root.chunk.js"
    ></script>
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  "
`;

exports[`seo ssr test default meta pack 1`] = `
"
    <meta charset="UTF-8" >
    
    <meta name="attrFromPageService" content="\${SERVER_URL}/seo/default/" data-meta-dynamic="true" >
    <meta name="metaCustomNameAttribute" content="metaCustomContent" data-meta-dynamic="true" >
    

    
    <script></script>
    

    
    <link crossorigin="anonymous"
      data-critical="true"
      onload="__preloadJS()"
      rel="stylesheet"
      href="\${STATIC_URL}/dist/client/root.chunk.css"
    >
    

    

    

    

    

    
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/root.chunk.js"
    ></script>
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  "
`;

exports[`seo ssr test json-ld meta 1`] = `
"
      {
        "@graph": [
          {
            "url": "https://www.tinkoff.ru/",
            "name": "Тинькофф Банк",
            "@type": "BankOrCreditUnion",
            "image": "https://static.tinkoff.ru/dist/portal-new/compiled/be23084191c7e7b4012da807d829ac04.png",
            "address": {
              "@type": "PostalAddress",
              "postalCode": "123060",
              "streetAddress": "1-й Волоколамский проезд, д.10, стр.1",
              "addressCountry": "Россия",
              "addressLocality": "Москва"
            },
            "@context": "http://schema.org/",
            "legalName": "АО «Тинькофф Банк»",
            "telephone": "+7 800 333-777-3",
            "priceRange": "0-7999",
            "description": "лицензия ЦБ РФ № 2673",
            "contactPoint": {
              "@type": "ContactPoint",
              "telephone": "+7 800 333-777-3",
              "areaServed": "Для звонков по России",
              "contactType": "customer service",
              "contactOption": "TollFree"
            }
          },
          {
            "url": "https://www.tinkoff.ru/",
            "name": "Тинькофф Банк",
            "@type": "Organization",
            "image": "https://static.tinkoff.ru/dist/portal-new/compiled/be23084191c7e7b4012da807d829ac04.png",
            "address": {
              "@type": "PostalAddress",
              "postalCode": "123060",
              "streetAddress": "1-й Волоколамский проезд, д.10, стр.1",
              "addressCountry": "Россия",
              "addressLocality": "Москва"
            },
            "@context": "http://schema.org/",
            "legalName": "АО «Тинькофф Банк»",
            "telephone": "+7 800 333-777-3",
            "description": "лицензия ЦБ РФ № 2673",
            "contactPoint": {
              "@type": "ContactPoint",
              "telephone": "+7 800 333-777-3",
              "areaServed": "Для звонков по России",
              "contactType": "customer service",
              "contactOption": "TollFree"
            },
            "aggregateRating": { "@type": "AggregateRating", "ratingValue": "5", "reviewCount": "439" }
          },
          {
            "name": "Кредитная карта «Тинькофф Платинум»",
            "@type": "Product",
            "image": {
              "url": "https://static2.tinkoff.ru/portal/mobile-home/img-platinum-char.png",
              "@type": "ImageObject"
            },
            "offers": {
              "@type": "AggregateOffer",
              "lowPrice": "590",
              "offeredBy": {
                "name": "Тинькофф Банк",
                "@type": "BankOrCreditUnion",
                "image": {
                  "url": "https://static.tinkoff.ru/dist/portal-new/compiled/c5e2e251a557ad1e3a5f120d5516d00d.png",
                  "@type": "ImageObject"
                },
                "address": {
                  "@type": "PostalAddress",
                  "postalCode": "123060",
                  "streetAddress": "1-й Волоколамский проезд, д.10, стр.1",
                  "addressCountry": "Россия",
                  "addressLocality": "Москва"
                },
                "telephone": "+7 800 555-777-8",
                "priceRange": "0-7999"
              },
              "priceCurrency": "RUB"
            },
            "@context": "http://schema.org",
            "description": "Кредитный лимит на карте до 300 000 рублей. Без процентов до 55 дней",
            "potentialAction": {
              "url": "https://www.tinkoff.ru/cards/credit-cards/tinkoff-platinum/#form",
              "name": "Оформить карту",
              "@type": "Action"
            }
          }
        ],
        "@context": "http://schema.org"
      }
    "
`;

exports[`seo ssr test og meta 1`] = `
"
    <meta charset="UTF-8" >
    
    <meta name="attrFromPageService" content="\${SERVER_URL}/seo/og/" data-meta-dynamic="true" >
    <title data-meta-dynamic="true">og seo</title>
    <meta property="og:description" content="og desc" data-meta-dynamic="true" >
    <meta property="og:image" content="og img" data-meta-dynamic="true" >
    <meta property="og:image:alt" content="alt og img" data-meta-dynamic="true" >
    <meta property="og:image:secure_url" content="seo imageSecure" data-meta-dynamic="true" >
    <meta property="og:image:height" content="seo 630" data-meta-dynamic="true" >
    <meta property="og:image:width" content="seo 1200" data-meta-dynamic="true" >
    <meta property="og:image:type" content="ggog" data-meta-dynamic="true" >
    <meta property="og:locale" content="locog" data-meta-dynamic="true" >
    <meta property="og:site_name" content="site og name" data-meta-dynamic="true" >
    <meta property="og:title" content="hello, og" data-meta-dynamic="true" >
    <meta property="og:type" content="og typ" data-meta-dynamic="true" >
    <meta property="og:url" content="http://og.og/" data-meta-dynamic="true" >
    <meta name="metaCustomNameAttribute" content="metaCustomContent" data-meta-dynamic="true" >
    

    
    <script></script>
    

    
    
    <style type="text/css"></style>
    <link crossorigin="anonymous" data-href="\${STATIC_URL}/dist/client/root.chunk.css" rel="stylesheet" >
    

    

    

    

    

    
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/root.chunk.js"
    ></script>
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  "
`;

exports[`seo ssr test twitter meta 1`] = `
"
    <meta charset="UTF-8" >
    
    <meta name="attrFromPageService" content="\${SERVER_URL}/seo/twitter/" data-meta-dynamic="true" >
    <title data-meta-dynamic="true">twitter seo</title>
    <meta name="twitter:card" content="twitter card" data-meta-dynamic="true" >
    <meta name="twitter:creator" content="creat" data-meta-dynamic="true" >
    <meta name="twitter:image" content="img tw" data-meta-dynamic="true" >
    <meta name="twitter:image:alt" content="seo tw im al" data-meta-dynamic="true" >
    <meta name="twitter:site" content="seo @tinkoff_bank" data-meta-dynamic="true" >
    <meta name="twitter:title" content="hello, twitter" data-meta-dynamic="true" >
    <meta name="metaCustomNameAttribute" content="metaCustomContent" data-meta-dynamic="true" >
    

    
    <script></script>
    

    
    
    <style type="text/css"></style>
    <link crossorigin="anonymous" data-href="\${STATIC_URL}/dist/client/root.chunk.css" rel="stylesheet" >
    

    

    

    

    

    
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/root.chunk.js"
    ></script>
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/react.js"
    ></script>
    <script defer="defer"
      charset="utf-8"
      crossorigin="anonymous"
      data-critical="true"
      src="\${STATIC_URL}/dist/client/platform.js"
    ></script>
    

    

    

    

    
  "
`;
